juju/utils.py

```diff
@@ -93,20 +95,31 @@ class IdQueue:
     ID.
     """
 
-    def __init__(self, maxsize=0):
-        self._queues = defaultdict(partial(jasyncio.Queue, maxsize))
-
-    async def get(self, id_):
+    _queues: dict[int, asyncio.Queue[dict[str, Any] | Exception]]
+
+    def __init__(self):
+        self._queues = defaultdict(asyncio.Queue)
+        # FIXME cleanup needed.
+        # in some cases an Exception is put into the queue.
+        # if the main coro exits, this exception will be logged as "never awaited"
+        # we gotta do something about that to keep the output clean.
+        #
+        # Additionally, it's conceivable that a response is put in the queue
+        # and then an exception is put via put_all()
+        # the reader only ever fetches one item, and exception is "never awaited"
+        # rewrite put_all to replace the pending response instead.
+
+    async def get(self, id_: int) -> dict[str, Any]:
         value = await self._queues[id_].get()
         del self._queues[id_]
         if isinstance(value, Exception):
             raise value
         return value
 
-    async def put(self, id_, value):
+    async def put(self, id_: int, value: dict[str, Any]):
         await self._queues[id_].put(value)
 
-    async def put_all(self, value):
+    async def put_all(self, value: Exception):
         for queue in self._queues.values():
             await queue.put(value)
```

Comment:
Interesting signatures, I won't pry too deeply here, I assume they just reflect how the methods are actually used.

But do you think it's worth annotating `-> None:` (in general) to make it clear that the return type has been considered?

Reply:
indeed these type hints are for semantically makes sense to call these methods with.

the implementation doesn't care, ofc.

I'm not sure about None return types. I think the annotation is useful only in a limited way -- prevents someone accidentally using the return value, or shows them not to in an IDE.